src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonArray.cs

```diff
@@ -180,34 +181,41 @@ public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? optio
 
         private void CreateNodes()
         {
-            if (_list == null)
+            // Even though _list initialization can be subject to races,
+            // ensure that contending threads use a coherent view of jsonElement.
+
+            JsonElement? jsonElement = _jsonElement;
+            Interlocked.MemoryBarrier();
+            List<JsonNode?>? list = _list;
+
+            if (list is null)
```

Comment:
Direct check of _list? `if (_list is null)`?

Reply:
It's possible, but I think making the read statements explicit more clearly convenes the fact that order of reads is significant.